openerp/addons/base/res/res_users.py

```diff
@@ -616,13 +616,39 @@ def create(self, cr, uid, values, context=None):
         return gid
 
     def write(self, cr, uid, ids, values, context=None):
+        """
+        Apply new implied groups to all users in the group, and apply all
+        implied groups to the new users in the group.
+        Only process proper additions by comparing the set of users and
+        implied groups with the set before the write.
+        """
+        if values.get('implied_ids'):
+            implied_groups_map = dict(
+                [(group.id, set([implied_id.id
+                                 for implied_id in group.trans_implied_ids]))
+                 for group in self.browse(cr, uid, ids, context=context)])
+        if values.get('users'):
+            users_map = dict(
+                [(group.id, set([user.id for user in group.users]))
+                 for group in self.browse(cr, uid, ids, context=context)])
         res = super(groups_implied, self).write(cr, uid, ids, values, context)
-        if values.get('users') or values.get('implied_ids'):
-            # add all implied groups (to all users of each group)
+        if values.get('implied_ids'):
+            # add all implied groups (to all users of each group) if necessary
+            for g in self.browse(cr, uid, ids, context=context):
+                if implied_groups_map[g.id] < set([
+                        implied_id.id
+                        for implied_id in g.trans_implied_ids]):
+                    gids = map(int, g.trans_implied_ids)
+                    vals = {'users': [(4, u.id) for u in g.users]}
+                    super(groups_implied, self).write(cr, uid, gids, vals, context)
+        if values.get('users'):
+            # add all implied groups to the new users
             for g in self.browse(cr, uid, ids, context=context):
-                gids = map(int, g.trans_implied_ids)
-                vals = {'users': [(4, u.id) for u in g.users]}
-                super(groups_implied, self).write(cr, uid, gids, vals, context)
+                new_users = set([user.id for user in g.users])
+                if users_map[g.id] < new_users:
```

Comment:
No, removing implied groups (or removing users from groups that have implied groups) are and were not processed because there is no way of knowing if the previously implied group wasn't assigned manually.